drivers/storage/class/class2/class2.c

```diff
@@ -4691,20 +4725,14 @@ Return Value:
         } else {
             deviceExtension->PhysicalDevice = deviceObject;
         }
+
+        deviceExtension->DeviceName = ntUnicodeString;
     }
 
     deviceObject->Flags &= ~DO_DEVICE_INITIALIZING;
 
     *DeviceObject = deviceObject;
 
-    RtlFreeUnicodeString(&ntUnicodeString);
```

Comment:
You now need to free this as part of device cleanup (i.e. IRP_MN_REMOVE_DEVICE or wherever IoDeleteDevice is called)

Reply:
There's no such possiblity. This function is called by class driver (look at disk for instance), but there's no such removal code (yet?)

Reply:
Hmm I see. Could you add the size of the buffer to the device extension size, and have the string buffer itself in there?

Reply:
Ah, that also comes from the caller. Yikes.